### DISCUSSION TITLE
PKG_CONFIG_PATH

### DISCUSSION THREAD

**Discussion by sekumare** [mesonbuild/meson#9977] [General]
Hi,

I am using meson version 0.55.1, I trying to cross compile libgudev package. 
I have dependecy glib-2.o is cross compiled and available in different directory.  I tried the following cross-file but nothing helped. 
Please let me know the correct way of using PKG_CONFIG_PATH 

pkg_config_path = [ ' <my path> ' ] - Not working
PKG_CONFIG_PATH = ' ' - Not working

**My cross-file for reference :**
---------------------------------

```ini
[binaries]
# we could set exe_wrapper = qemu-arm-static but to test the case
# when cross compiled binaries can't be run we don't do that
c = '/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-gcc'
cpp = '/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-g++'
rust = ['rustc', '--target', 'arm-buildroot-linux-gnueabi', '-C', ]
ar = '/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-ar'
strip = '/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-strip'
pkgconfig = '/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/lib/pkgconfig'
ld = '/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-ld'
cmake = 'cmake'

[built-in options]
# Used in unit test '140 get define'
c_args = ['-DMESON_TEST_ISSUE_1665=1']
cpp_args = '-DMESON_TEST_ISSUE_1665=1'

[properties]
has_function_printf = true
has_function_hfkerhisadf = false
skip_sanity_check = true
c_args = []
cpp_args = []
objc_args = []
objcpp_args = []
c_link_args = ['-L/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32//lib', '-Wl,-rpath-link=/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32//lib']
cpp_link_args = ['-L/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32//lib', '-Wl,-rpath-link=/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32//lib']
objc_link_args = ['-L/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32//lib', '-Wl,-rpath-link=/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32//lib']
objcpp_link_args = ['-L/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32//lib', '-Wl,-rpath-link=/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32//lib']
pkg_config_libdir = ['/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32//lib/pkgconfig:/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32//usr/share/pkgconfig']
sys_root = '/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/'


[host_machine]
system = 'linux'
cpu_family = 'arm'
cpu = 'armv7' # Not sure if correct.
endian = 'little'
```

**Comment by dcbaker** [comment]
Check the contents of your pkg-config wrapper, a lot of them do something like this:
```sh
#!/bin/sh
PKG_CONFIG_PATH=<some set of cross paths> pkg-config $@
```
Which means that any changes you make to PKG_CONFIG_PATH are ignored by the wrapper

**Reply by sekumare** [reply]
Thanks. I did check the libgudev package meson. build file.  I am not sure where this 'pkgconfig' variable is taken from.  Any way to override in the cross-file ?. 

```meson
gnome = import('gnome')
pkg = import('pkgconfig')

# Versioning
api_version = '1.0'
lib_version = '0.3.0'
version = meson.project_version()

package_name = 'gudev'
package_string = '@0@-@1@'.format(package_name, api_version)

# Paths
rootdir = include_directories('.')

srcdir = meson.source_root()
builddir = meson.build_root()

prefix = get_option('prefix')

includedir = join_paths(prefix, get_option('includedir'))
libdir = join_paths(prefix, get_option('libdir'))
datadir = join_paths(prefix, get_option('datadir'))

pkgincludedir = join_paths(includedir, package_string)
typelibdir = join_paths(libdir, 'girepository-1.0')
pkgconfigdir = join_paths(libdir, 'pkgconfig')
girdir = join_paths(datadir, 'gir-1.0')
gtkdocdir = join_paths(datadir, 'gtk-doc')
vapidir = join_paths(datadir, 'vala', 'vapi')

# Dependencies

glib_req = '>= 2.38.0'
libudev_req = '>= 199'
introspection_req = '>= 1.31.1'
vapigen_req = '>= 0.38.0'
gtk_doc_req = '>= 1.18'
```

**Reply by dcbaker** [reply]
It's this line in your cross file:
```
pkgconfig = '/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/lib/pkgconfig'
```
if you run:
`cat /opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/lib/pkgconfig`
what does it return?

**Comment by sekumare** [comment]
It is a directory that containse list of package config files.
```txt
# ll /opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/lib/pkgconfig
total 32
-rw-r--r--. 1 14852 users 296 Feb 11  2020 expat.pc
-rw-r--r--. 1 14852 users 531 Feb 11  2020 isl.pc
-rw-r--r--. 1 14852 users 294 Feb 11  2020 libacl.pc
-rw-r--r--. 1 14852 users 307 Feb 11  2020 libattr.pc
-rw-r--r--. 1 14852 users 381 Feb 11  2020 libpkgconf.pc
-rw-r--r--. 1 14852 users 342 Feb 11  2020 mpfr.pc
-rw-r--r--. 1 14852 users 328 Feb 11  2020 python-2.7.pc
lrwxrwxrwx. 1 14852 users  13 Feb 11  2020 python2.pc -> python-2.7.pc
lrwxrwxrwx. 1 14852 users  10 Feb 11  2020 python.pc -> python2.pc
-rw-r--r--. 1 14852 users 323 Feb 11  2020 zlib.pc
```

But glib-2.0 (dependency library) is in the below directory
/home/mylibs/glib/glib-2.56.4/glib/

I have passed that also in my crossfile but always gettign any error 

WARNING: The version of CMake /usr/bin/cmake is 2.8.12.2 but version >=3.4 is required
Run-time dependency glib-2.0 found: NO

meson.build:49:0: ERROR: Pkg-config binary for machine MachineChoice.HOST not found. Giving up.

**Reply by dcbaker** [reply]
ah, that should point at a `pkg-config` executable (either a binary or a script), not to a directory

If `/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-pkg-config` exists, you probably want to use that, otherwise you can write one like this:

```sh
#!/bin/sh

PKG_CONFIG_LIBDIR=/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/lib/pkgconfig pkg-config $@
```
then make that executable and point to that in your cross file

**Comment by eli-schwartz** [comment]
```ini
[binaries]
# ...
pkgconfig = '/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/lib/pkgconfig'
```

This is incredibly suspicious, because based on the basic structure of that path it is almost guaranteed to be a sysroot *directory* of `*.pc` files.

However, you told the cross file this is the location of a *binary*, specifically, this is the location of the pkg-config *program*. So, Meson tried to use this directory when looking up the program, and got a "cannot execute file: is a directory" type of error, so as a result it said it cannot find pkg-config at all.

You did *not* pass this directory as you should have done, to the *property* `pkg_config_libdir`: https://mesonbuild.com/Cross-compilation.html#properties

**Comment by sekumare** [comment]
Hi Eli,

I just trimmed my cross file a bit and added the pkg lib conf directory as
per the suggested link.  The issue still persists.

```ini
[binaries]
# we could set exe_wrapper = qemu-arm-static but to test the case
# when cross compiled binaries can't be run we don't do that
c =
'/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-gcc'
cpp =
'/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-g++'
rust = ['rustc', '--target', 'arm-buildroot-linux-gnueabi', '-C', ]
ar =
'/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-ar'
strip =
'/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-strip'
ld =
'/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-ld'
cmake = 'cmake'

[built-in options]
# Used in unit test '140 get define'
c_args = ['-DMESON_TEST_ISSUE_1665=1' ]
cpp_args = '-DMESON_TEST_ISSUE_1665=1'

[properties]
has_function_printf = true
has_function_hfkerhisadf = false
skip_sanity_check = true
sys_root =
'/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/'
pkg_config_libdir = '/home/mylibs/glib/glib-2.56.4/glib/'


[host_machine]
system = 'linux'
cpu_family = 'arm'
cpu = 'armv7' # Not sure if correct.
endian = 'little'
```

*Error : *

C compiler for the build machine: cc (gcc 7.3.1 "cc (GCC) 7.3.1 20180303
(Red Hat 7.3.1-5)")
C linker for the build machine: cc ld.bfd 2.28-11
C compiler for the host machine:
/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-gcc
(gcc 9.2.0 "arm-buildroot-linux-gnueabi-gcc.br_real (Buildroot 2019.11.1)
9.2.0")
C linker for the host machine:
/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-gcc
ld.bfd 2.32
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: arm
Host machine cpu: armv7
Target machine cpu family: arm
Target machine cpu: armv7
Found Pkg-config: NO
Found CMake: /usr/bin/cmake (2.8.12.2)
WARNING: The version of CMake /usr/bin/cmake is 2.8.12.2 but version >=3.4
is required
*Run-time dependency glib-2.0 found: NO*

**Reply by dcbaker** [reply]
you still need to point to *some* pkg-config, just `pkg-config =/usr/bin/pkg-config` should work if you have `pkg_config_libdir` set.

**Comment by sekumare** [comment]
Thanks for your input. It worked. It moved ahead and now it is expecting
another udev version. I will debug further.

 compiler for the build machine: cc (gcc 7.3.1 "cc (GCC) 7.3.1 20180303
(Red Hat 7.3.1-5)")
C linker for the build machine: cc ld.bfd 2.28-11
C compiler for the host machine:
/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-gcc
(gcc 9.2.0 "arm-buildroot-linux-gnueabi-gcc.br_real (Buildroot 2019.11.1)
9.2.0")
C linker for the host machine:
/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-gcc
ld.bfd 2.32
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: arm
Host machine cpu: armv7
Target machine cpu family: arm
Target machine cpu: armv7
Found pkg-config:
/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/pkg-config
(1.6.1)
Run-time dependency glib-2.0 found: YES 2.56.4
Run-time dependency gobject-2.0 found: YES 2.56.4
Dependency libudev found: NO found 136 but need: '>= 199'
Found CMake: /usr/bin/cmake (2.8.12.2)
WARNING: The version of CMake /usr/bin/cmake is 2.8.12.2 but version >=3.4
is required
Run-time dependency libudev found: NO

meson.build:51:0: ERROR: Invalid version of dependency, need 'libudev' ['>=
199'] found '136'.

**Comment by sekumare** [comment]
Thanks for your input. It worked. It moved ahead and now it is expecting
another udev version. I will debug further.

 compiler for the build machine: cc (gcc 7.3.1 "cc (GCC) 7.3.1 20180303
(Red Hat 7.3.1-5)")
C linker for the build machine: cc ld.bfd 2.28-11
C compiler for the host machine:
/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-gcc
(gcc 9.2.0 "arm-buildroot-linux-gnueabi-gcc.br_real (Buildroot 2019.11.1)
9.2.0")
C linker for the host machine:
/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-gcc
ld.bfd 2.32
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: arm
Host machine cpu: armv7
Target machine cpu family: arm
Target machine cpu: armv7
Found pkg-config:
/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/pkg-config
(1.6.1)
Run-time dependency glib-2.0 found: YES 2.56.4
Run-time dependency gobject-2.0 found: YES 2.56.4
Dependency libudev found: NO found 136 but need: '>= 199'
Found CMake: /usr/bin/cmake (2.8.12.2)
WARNING: The version of CMake /usr/bin/cmake is 2.8.12.2 but version >=3.4
is required
Run-time dependency libudev found: NO

meson.build:51:0: ERROR: Invalid version of dependency, need 'libudev' ['>=
199'] found '136'.

**Comment by sekumare** [comment]
Today, I have tried another systemd package using meson. from the meson
configure logs, the following is my output. However, during ninja build, my
cross compilation tool path is getting appended to my source path and
complaining -Wmissing including header and I am not sure how the cross tool
chain path is appended.

*Note : libmount and libblk are present. *

Message: systemd 243
         split /usr:                        false
         split bin-sbin:                    true
         prefix directory:                  /home/mypath/fs.build
         rootprefix directory:              /usr
         sysconf directory:                 /etc
         include directory:                /home/mypath/fs.build/include
         lib directory:                      /home/build/fs.build/lib
         rootlib directory:                 /usr/lib
         SysV init scripts:                 /etc/init.d
         SysV rc?.d directories:            /etc/rc.d
         PAM modules directory:             /usr/lib/security
         PAM configuration directory:       /etc/pam.d

Build targets in project: 1579
Option prefix is: /home/mypath/fs.build [default: /usr]
Found ninja-1.10.2 at /usr/bin/ninja
mkdir -p /home/mypath/fs.build/public/systemd
cd objs/arm; DESTDIR=/home/mypath/fs.build/public/systemd
/home/mypath/ninja/ninja-1.9.0/ninja install
[14/1437] Compiling C object 'src/core/2ac6ece@@***@***.***/kill.c.o'.
cc1: warning:
/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/home/mypath/fs.build/public/include/libmount:
No such file or directory [-Wmissing-include-dirs]
cc1: warning:
/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/home/mypath/fs.build/public/include/blkid:
No such file or directory [-Wmissing-include-dirs]
[36/1437] Compiling C object 'src/basic/a6ba3eb@@***@***.***/MurmurHash2.c.o'.


*Cross file  :*

[binaries]
# we could set exe_wrapper = qemu-arm-static but to test the case
# when cross compiled binaries can't be run we don't do that
c =
'/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-gcc'
cpp =
'/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-g++'
rust = ['rustc', '--target', 'arm-buildroot-linux-gnueabi', '-C', ]
ar =
'/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-ar'
strip =
'/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-strip'
pkgconfig =
'/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/pkg-config'
ld =
'/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/bin/arm-buildroot-linux-gnueabi-ld'
cmake = 'cmake'

[built-in options]
# Used in unit test '140 get define'
c_args = ['-DMESON_TEST_ISSUE_1665=1' ]
cpp_args = '-DMESON_TEST_ISSUE_1665=1'

[properties]
has_function_printf = true
has_function_hfkerhisadf = false
skip_sanity_check = true
sys_root =
'/opt/toolchains/crosstools-arm-gcc-9.2-linux-4.19-glibc-2.30-binutils-2.32/'
pkg_config_libdir = [ '/home/mypath/fs.build/lib/pkgconfig/'  ]


[host_machine]
system = 'linux'
cpu_family = 'arm'
cpu = 'armv7' # Not sure if correct.
endian = 'little'

~
~
~
~

**Comment by sekumare** [comment]
My meson options :
SYSTEMD_MESON_OPTIONS=-Db_pch=false -Dprefix=$(mypath)
-Dc_args='-I$(mypath)/public/include -I$(mypath)/public/usr/include
-I$(CURR_DIR)/$(APP)' -Dc_link_args='-L$(mypath)/public/lib
-L$(BCM_FSBUILD_DIR)/gpl/lib -L$(INSTALL_DIR)/lib'

SYSTEMD_PROJECT_OPTIONS=-Dpam=false -Dkmod=false -Dselinux=false
-Dapparmor=false -Dadm-group=false -Dwheel-group=false -Dxz=false
-Dzlib=false -Dlibidn=false -Dlibiptc=false -Dgcrypt=false
-Dxkbcommon=false -Ddbus=false -Dglib=false

**Comment by sekumare** [comment]
The issue is my package required "Run time dependency of  libpolkit-gobject
<https://freedesktop.org/wiki/Software/polkit>".

Is the anyway to disable the above package. I did attempt
-Denable_polkit=no but that doesnt help.

Thanks,
SK
